partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/Scope.kt

```diff
@@ -62,6 +65,30 @@ internal data class Scope(
      */
     override fun toString(): String = "< " + schema.joinToString { "${it.name}: ${it.type}" } + " >"
 
+    /**
+     * Check if qualified identifier matches a binding with qualifier and name.
+     */
+    private fun matchQualified(identifier: Identifier, depth: Int = 0): Rex? {
+        val path = identifier.toList()
+        val qualifierParts = path.dropLast(1)
+        val bindingName = path.last()
+        for (i in schema.indices) {
+            val local = schema[i]
+            val type = local.type
+            if (local.qualifier != null && bindingName.matches(local.name)) {
+                val localQualifierParts = local.qualifier.split(".")
+                if (qualifierParts.size == localQualifierParts.size &&
+                    qualifierParts.zip(localQualifierParts).all { (identPart, localPart) ->
```

Comment:
This is a pretty complicated nesting of `if`s and conditionals. Consider inverting the `if`s to make it more readable -- https://stackoverflow.com/questions/268132/why-should-i-invert-if-statement-to-reduce-nesting.